src/app/Media/VoIPMediaSession.cs

```diff
@@ -206,42 +222,56 @@ private void VideoFormatsNegotiated(List<VideoFormat> videoFormats)
             _videoTestPatternSource?.SetVideoSourceFormat(videoFormat);
         }
 
+        private void TextFormatsNegotiated(List<TextFormat> textFormats)
+        {
+            var textFormat = textFormats.First();
+            logger.LogDebug($"Setting text sink and source format to {textFormat.FormatID} {textFormat.FormatName}. (RTP clock rate {textFormat.ClockRate}).");
+            Media.TextSource?.SetTextSourceFormat(textFormat);
+        }
+
         public async override Task Start()
         {
-            if (!base.IsStarted)
+            await base.Start().ConfigureAwait(false);
+
+            if (HasAudio)
             {
-                await base.Start().ConfigureAwait(false);
+                if (Media.AudioSource != null)
+                {
+                    await Media.AudioSource.StartAudio().ConfigureAwait(false);
+                }
+                if (Media.AudioSink != null)
+                {
+                    await Media.AudioSink.StartAudioSink().ConfigureAwait(false);
+                }
+            }
 
-                if (HasAudio)
+            if (HasVideo && Media.VideoSource != null)
+            {
+                if (!_videoCaptureDeviceFailed)
                 {
-                    if (Media.AudioSource != null)
-                    {
-                        await Media.AudioSource.StartAudio().ConfigureAwait(false);
-                    }
-                    if (Media.AudioSink != null)
-                    {
-                        await Media.AudioSink.StartAudioSink().ConfigureAwait(false);
-                    }
+                    await Media.VideoSource.StartVideo().ConfigureAwait(false);
                 }
+                else
+                {
+                    logger.LogWarning($"Webcam video source failed before start, switching to test pattern source.");

```

Comment:
As per #1306, use structured semantic logging instead of dynamic log message templates.